### PR TITLE
fix: render CommonJS externals directly in modern-module output

### PR DESCRIPTION
**Pull request by JSerFeng**
## Summary

- Render eligible static CommonJS externals directly at their CommonJS dependency sites for `library.type: "modern-module"`, removing the intermediate external module factory.
- Preserve wrapper semantics for optional, conditional/context, request-rest, relative, and base-sensitive `node-commonjs` requests, including reused module graphs and mixed/multi-entry consumers.
- Handle Node builtin export-requires and object-form external maps safely, while sharing the collision-safe `createRequire` init fragment.

For an eligible external, emitted code now keeps `require("external")` (or the `createRequire` equivalent for Node builtins) in the rendered module instead of routing through a generated `external "external"` factory.

## Related links

Supersedes #14948.

## Validation

- `corepack pnpm run build:binding:dev`
- `corepack pnpm run build:js`
- `corepack pnpm test EsmOutput.test.js` (522 tests)
- focused modern-module concatenation and external edge-case suites
- `cargo test -p rspack_plugin_esm_library commonjs_external::tests`
- `cargo test -p rspack_core module_graph::connection::tests`
- `cargo clippy -p rspack_core -p rspack_plugin_esm_library -p rspack_plugin_externals --all-targets -- -D warnings`
- `cargo fmt --all --check`
- `corepack pnpm run lint:js`

## Checklist

- [x] Tests updated.
- [x] Documentation not required (internal rendering optimization; no public API change).

### CODE REVIEW

**Comment by alfaproject** [comment]
Oh man, this has been a major pet peeve of mine, I'm glad it's being worked on. Also duplicate imports from `node:path` and `path` is another one

**Comment by JSerFeng** [comment]
@codex review

File: crates/rspack_plugin_esm_library/src/dependency/commonjs_external.rs [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+      else {
+        continue;
+      };
+
+      let request = external_module.get_request();
+      if CommonJsExternalRequireKind::from_external_type(external_module.resolve_external_type())
+        .is_some()
+        && !request.has_rest()
+        // Relative calls must stay in the external module's emitted directory.
+        && !is_relative_external_request(request.primary())
```
[the remark this answers is not in the document]
Comment (JSerFeng) [reply crates/rspack_plugin_esm_library/src/dependency/commonjs_external.rs:124]: It has same issue all along, actually, chunking is controlled by user, so we cannot guarrenting relative path resolving here

**Comment by JSerFeng** [comment]
@codex review

**Comment by JSerFeng** [comment]
> ### 💡 Codex Review
> https://github.com/web-infra-dev/rspack/blob/83a28c01adf2a564d02e579dcf46259ee6503e66/crates/rspack_plugin_esm_library/src/plugin.rs#L811-L815
> 
> 
> **P1 Badge Preserve partially used exports when merging externals**
> When two issuer-scoped `module` externals are merged into one chunk and a non-canonical copy uses only a nested property (for example, one issuer consumes `external.a.x`), its top-level export is marked `OnlyPropertiesUsed`, not `Used`. This filter therefore omits that export from `used_exports`; after all connections are moved to the canonical module, usage analysis is not rerun, so `collect_module_external_remapping` can treat the export as unused and generate a namespace without it. Merge `OnlyPropertiesUsed` and its nested usage state, or conservatively promote every non-`Unused` export to `Used` on the canonical module.
> 
> ℹ️ About Codex in GitHub

It's the same with current main, as external module cannot produce nested exports info

**Review by Timeless0911 — APPROVED** [review]

**Comment by SoonIter** [comment]
<agent: daily-job rspack eco-ci>

### Eco-CI: `rspack/rslib` expects the old external wrapper

**Conclusion / impact:** This PR exposes three downstream output-shape assertions; the focused runtime checks pass after adjusting them. Rslib repair: [#1931](https://github.com/web-infra-dev/rslib/pull/1931).

#### Why this PR
The modern-module dependency templates now emit CommonJS external calls directly. Rslib still expects `module.exports = __rspack_createRequire_require(...)`. The same Rslib SHA passes before this merge and fails at the merge commit; generated configs enable modern-module. This is a downstream test-contract mismatch, not evidence of a runtime regression.

#### Suggested fix
Rslib should assert the external callee/request and retain its runtime checks without requiring the removed wrapper. The repair passes all 10 externals tests with both the locked dependencies and this PR's official WASM artifact.

<details>
<summary>Evidence and attribution checks</summary>

- Current run: https://github.com/rstackjs/rstack-ecosystem-ci/actions/runs/35323623709
- Current job: https://github.com/rstackjs/rstack-ecosystem-ci/actions/runs/35323623709/job/105531676228
- Tested Rspack: `3861cb025c648c93760fd09f0d32587dc55872d8`; downstream Rslib: `57087f50a80fecfc706e1d296444432cce9eab3a`.
- Previous main green: run 35215608792 / job 105183339567; first bad: run 35304243186 / job 105473133524 at `475c7ad20f5dded5c1139e3bc820673c9a59f015`. Both use that same Rslib SHA.
- Local comparison: official WASM artifact from Rspack CI run 35303892819, core JS 2.2.6 (unchanged by this PR): original assertions 3 failed / 7 passed; repaired assertions 10 passed. Stable locked Rspack 2.2.4: 10 passed.
- `DEBUG=rsbuild` fixture configs confirm `output.library.type` and `externalsType` are `modern-module`.
- Latest surface #15758 changes only Rspack tests. Interleaved migration run 35322823195 does not contain this PR and is not a flaky counterexample. The signature is absent from prior green main runs and reproduced deterministically with this PR's artifact.
- Full latest Linux-native suite was not rerun locally. No dependency upgrades or runtime changes are proposed.

</details>
